include/revng/Yield/Support/GraphLayout/SugiyamaStyle/InternalGraph.h

```diff
@@ -0,0 +1,234 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+#include <variant>
+
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Yield/Support/GraphLayout/Traits.h"
+
+namespace yield::layout::sugiyama {
+
+class InternalGraph;
+
+namespace detail {
+
+struct InternalNode {
+  layout::Point Center;
+  layout::Size Size;
+
+  bool IsVirtual;
+
+private:
+  std::size_t Index;
+
+private:
+  friend InternalGraph;
+  InternalNode(std::size_t Index, layout::Size &&Size) :
+    Center({ 0, 0 }), Size(std::move(Size)), IsVirtual(false), Index(Index) {}
+  InternalNode(std::size_t Index) :
+    Center({ 0, 0 }), Size({ 0, 0 }), IsVirtual(true), Index(Index) {}
+
+public:
+  InternalNode(const InternalNode &) = delete;
+  InternalNode(InternalNode &&) = default;
+  InternalNode &operator=(InternalNode &&) = default;
+
+  std::size_t index() const { return Index; }
+};
+
+struct InternalEdge {
+  bool IsRouted;
+  bool IsBackwards;
+
+private:
+  std::size_t Index;
+
+  using PathOwnership = std::unique_ptr<layout::Path>;
+  using PathReference = layout::Path *;
+  std::variant<PathOwnership, PathReference> Path;
+
+private:
+  // Only allow `InternalGraph` to construct these objects.
+  // This helps to keep the indexing consistent.
+  friend InternalGraph;
```

Comment:
Why are they all friends? I'm very suspicious about `friend` here. Moving the declarations of `InternalNode` and `InternalEdge` inside `InternalGraph` is better.

Reply:
Ok, strictly speaking you're right. At this point I'd consider one of the two following alternatives:
- if you can drop all the `friends` alltogether, leaving the nest of type declarations as they are now, I think that's the best option. At the moment I haven't tried, but what fails if you drop the `friend`?
- otherwise I think you can figure out something along with my comment on `ResultType` below

What do you think?

Reply:
Dropping `friend`ship would mean having to make the constructors public.

I guess it's not THAT big of a deal that something named `InternalX` is constructed in non-obvious way, so I'm not totally opposed, but it does feel wrong to let anyone at any point say "make me a new node with this arbitrary index I chose", especially considering how much care went into making sure no two nodes ever get the same index.

Reply:
Yeah no biggie in making `public` the constructors for the `InternalX`. It's not that they're are really public.

If you want to add some safeguards against misuse I think you can move the associated headers away from `include` and move them in `lib/Support/GraphLayout/SugiyamaStyle` (or something closer to that point). This effectively makes them "public" but only in that part of the code where it makes sense and where you really need to know what you're doing to use them. Or is there something preventing you to do that?

Reply:
Yep, the `make` and `exportInto` methods of the `InternalGraph` have to be templates. And they have to be able to access the class they create and dismantle.

Reply:
> Yeah no biggie in making `public` the constructors for the `InternalX`. It's not that they're are really public.

Actually it's a lot bigger than it seems. Because when they are `public`, they are constructible using the "natural" means, as in `InternalGraph.addNode()`. Which complains about not specifying index, BUT, if someone were to call `InternalGraph.addNode(0)` for some reason, that would probably be quite an issue to debug.

(for reference, the "canonical" ways to create nodes are `InternalGraph.makeNode()` and `InternalGraph.makeVirtualNode()`).

I guess it's not really sustainable as is - so we should probably leave this as a comment somewhere and/or also rename `makeX` things into something more natural.

There's also an option of preventing the constructor from being called using a `tag`-type in the constructor. But at that point, I'd rather just declare friendship and be done with it.

Reply:
How about having the constructor of the node take a reference to the `InternalGraph` you want to insert it into?

Then you can have:
- have a public method on `InternalGraph` that gives you the next available ID and increments the current ID: `getAndIncrementID() { return NodeIndexCounter++; }`
- All the constructors of `InternalNode` have to take an additional `InternalGraph &` argument. Something along this line: `InternalNode(InternalGraph &G) : ID(G.getAndIncrementID) { /* do stuff if necessary */ }` (I haven't specified `Size` explicitly, but you get the idea). In this way you cannot construct a node if you're not holding a graph, which makes it pretty obvious that the node should be part of that graph.
- The constructors for `InternalNode` that explicitly take the ID as parameter have to go away, so that nobody from outside can mess up the invariant that indices are unique.
- `InternalGraph::makeNode()` don't call the node constructors specifying the Index (since they are gone) but they use `InternalNode(*this)` (where `this` points to the `InternalGraph` you are inserting them to). What's nice is that this can now be done without them being `friend`s.
- So everything else is left unchanged, except that now someone can construct a `InternalNode` in the wild, as long as it has a `InternalGraph &` to provide to the constructor. This doesn't really strike me like something that can be heavily misused, and even if it was misused it would be very suspicious both for who writes it and for who reviews it.

Then something similar can be done for edges, and nobody is `friend` of anyone else anymore.

Reply:
Why haven't I thought of this! Yeah, this is definitely the way to go.

Reply:
Please take a fresh look. I'm really happy about how this turned out! Thanks to easier to access constructor, I was also able to get rid of `makeNode` and `makeEdge` outright! I still decided to keep the "virtual" counterparts though - I really like how explicit
```
auto *Node = Graph.makeVirtualNode();
// as opposed to
auto *Node = Graph.addNode(Graph);
```

Reply:
Nice! No `friend`s anymore. And it brought improvements also in adjacent parts of the codebase!